src/test/java/com/example/bookshop/controller/BookControllerTests.java

```diff
@@ -0,0 +1,243 @@
+                .andReturn();
+        List<BookDto> actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                objectMapper.getTypeFactory().constructCollectionType(List.class, BookDto.class));
+
+        assertNotNull(actual);
+        assertFalse(actual.isEmpty());
+        assertIterableEquals(
+                expected.stream().map(BookDto::getTitle).toList(),
+                actual.stream().map(BookDto::getTitle).toList()
+        );
+        assertIterableEquals(
+                expected.stream().map(BookDto::getAuthor).toList(),
+                actual.stream().map(BookDto::getAuthor).toList()
+        );
+    }
+
+    @Test
+    @DisplayName("GET /books/{id} should return single BookDto")
+    @WithMockUser
+    void getBookById_ReturnBookDto() throws Exception {
+        MvcResult result = mockMvc.perform(get("/books/{id}", 1L))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+
+        assertNotNull(actual);
+        assertEquals("Book 1", actual.getTitle());
+        assertEquals("Author 1", actual.getAuthor());
+    }
+
+    @Test
+    @DisplayName("GET /books/{id} when book doesn't exist should return 404")
+    @WithMockUser
+    void getBookById_NotFound() throws Exception {
+        mockMvc.perform(get("/books/{id}", 9999L))
+                .andExpect(status().isNotFound());
+    }
+
+    @Test
+    @DisplayName("GET /books/search should return filtered list (no params -> not empty)")
+    @WithMockUser
+    void searchBook_ReturnBookDtoList() throws Exception {
+        MvcResult result = mockMvc.perform(get("/books/search"))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        List<BookDto> actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(),
+                objectMapper.getTypeFactory().constructCollectionType(List.class, BookDto.class)
+        );
+
+        assertNotNull(actual);
+        assertFalse(actual.isEmpty());
+    }
+
+    @Test
+    @DisplayName("POST /books should create a book (ROLE_ADMIN)")
+    @WithMockUser(roles = "ADMIN")
+    void createBook_ReturnBookDto() throws Exception {
+        CreateBookRequestDto request = TestUtil.createBookRequestDto(BigDecimal.valueOf(99.99));
+        String json = objectMapper.writeValueAsString(request);
+
+        MvcResult result = mockMvc.perform(post("/books")
+                        .with(csrf())
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(json))
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        BookDto created = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+
+        assertNotNull(created);
+        assertEquals(request.getTitle(), created.getTitle());
```

Comment:
rename to expected and actual
```suggestion
        assertEquals(request, created);
```

Reply:
https://github.com/BelitskyiAlexandr/spring-boot-intro/pull/15#discussion_r2319770080

Reply:
In order for this to work, we need to override hashcode and equals for the DTO, although I don't see much point in this. assertEquals is a reference-based method, and here it is obvious that the objects will have different references.

Reply:
it's bad practice to compare objects by fields
please fix comment